[CALCITE-4059] SqlTypeUtil#equalSansNullability consider Array/Map type

Aaaaaaron opened the pull request.

Aaaaaaron commented:
Can someone review this PR?

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1151,13 +1151,22 @@ public static boolean equalSansNullability(
+    if (isArray(type1) && isArray(type2)) {
+      return equalSansNullability(factory, type1.getComponentType(), type2.getComponentType());
+    } else if (isMap(type1) && isMap(type2)) {
+      MapSqlType mType1 = (MapSqlType) type1;
```
I would suggest to add two new methods `equalAsArraySansNullability` and `equalsAsMapSansNullability`, because for most of the cases when we invoke these methods, we already know if the type to compare is a map/array/struct, add new methods simplify the original impl(which is the most common case).

danny0405 replied:
BTW, can you share why you need to compare map and array types ? Seems the case is rare from my side.

Aaaaaaron replied:
> I would suggest to add two new methods `equalAsArraySansNullability` and `equalsAsMapSansNullability`, because for most of the cases when we invoke these methods, we already know if the type to compare is a map/array/struct, add new methods simplify the original impl(which is the most common case).

Thanks for your advice, I'll change this latter.

Aaaaaaron replied:
> BTW, can you share why you need to compare map and array types ? Seems the case is rare from my side.

A case like this: array(12L) will be recognized as "INTEGER ARRAY, NOT NULL", but the col i_ids is "BIGINT ARRAY NULLABLE", so calcite will add a cast. But the source/targe nullability is still different, and we don't want to consider nullability.

```
INSERT TABLE new_t
SELECT id,
       array(12) AS i_ids,
FROM t
```

danny0405 replied:
Reasonable, in `TypeCoercionImpl#querySourceCoercion` we may need to take the array element nullability into account.

amaliujia commented:
Since you are adding support on this area, maybe also add other types's support (e.g. MultisetSqlType)?

Aaaaaaron commented:
> Since you are adding support on this area, maybe also add other types's support (e.g. MultisetSqlType)?

I'll take a look latter.

Aaaaaaron commented:
@danny0405 @amaliujia I've changed the PR, and split into two methods: :
 1. equalAsCollectionSansNullability, handles ARRAY/MULTISET
 2. equalAsMapSansNullability, handles MAP

amaliujia commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1510,6 +1563,31 @@ public static boolean isMap(RelDataType type) {
+  /**
+   * @return true if type is MULTISET
+   */
+  public static boolean isMultiset(RelDataType type) {
```
I think this function is not used as you are switching to `isCollection`

Aaaaaaron replied:
Indeed, I did not use this method, but basically, all RelDataType have a similar utility method, so I added this for the future use. If you think this is unnecessary, I can remove this.

amaliujia replied:
For me keeping this is ok.

Aaaaaaron replied:
![image](https://user-images.githubusercontent.com/15643702/84861572-ed8fbf00-b0a3-11ea-975a-4cf0ee1a1067.png)

amaliujia commented:
+1

LGTM

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1160,6 +1160,59 @@ public static boolean equalSansNullability(
+    }
+
+    if (isCollection(type1) && isCollection(type2)) {
+      return equalSansNullability(factory, type1.getComponentType(), type2.getComponentType());
```
And should a `MULTISET` equals to `ARRAY` ? I would say no.

amaliujia replied:
This is a good point. `type1. getSqlTypeName() == type2. getSqlTypeName()` is missing.

Aaaaaaron replied:
> we should add assertion instead in the first line: `assert isCollection(type1) && isCollection(type2)`

I'll add an assertion. Cuz assert is not enforce in Java, so I still keep the condition in if clause.

Aaaaaaron replied:
> And should a `MULTISET` equals to `ARRAY` ? I would say no.

Good point, that's a bug, thanks for your review, I've updated my PR.

danny0405 replied:
No need to have `isCollection(type1) && isCollection(type2)` which is already in the assertion.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1160,6 +1160,59 @@ public static boolean equalSansNullability(
 
+  /**
+   * Returns whether two array types are equal, ignoring nullability.
+   *
```
`array/multiset` -> `collection`

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1160,6 +1160,64 @@ public static boolean equalSansNullability(
+
+    if (type1.equals(type2)) {
+      return true;
+    }
```
Remove it.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1160,6 +1160,64 @@ public static boolean equalSansNullability(
+      return equalSansNullability(factory, mType1.getKeyType(), mType2.getKeyType())
+          && equalSansNullability(factory, mType1.getValueType(), mType2.getValueType());
+    } else {
+      return false;
```
`return equalSansNullability(factory, mType1.getKeyType(), mType2.getKeyType())
          && equalSansNullability(factory, mType1.getValueType(), mType2.getValueType());` is enough.

danny0405 commented on core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:
```diff
@@ -117,6 +120,47 @@
+  @Test void testEqualAsMapSansNullability() {
+    final SqlTypeFactoryImpl factory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+    final RelDataType bigint = factory.createSqlType(SqlTypeName.BIGINT);
```
Use type from `SqlTypeFixture` if possible.

Aaaaaaron replied:
> Use type from `SqlTypeFixture` if possible.

ok

Aaaaaaron replied:
Thanks for your suggestion, updated already.

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1160,6 +1161,49 @@ public static boolean equalSansNullability(
+      RelDataTypeFactory factory,
+      RelDataType type1,
+      RelDataType type2) {
+    Preconditions.checkArgument(isCollection(type1));
```
Add `type1 == type2` check to see if they are equals for fast check.

Aaaaaaron replied:
> Add `type1 == type2` check to see if they are equals for fast check.

ok

danny0405 commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1160,6 +1161,49 @@ public static boolean equalSansNullability(
+      RelDataTypeFactory factory,
+      RelDataType type1,
+      RelDataType type2) {
+    Preconditions.checkArgument(isMap(type1));
```
Add `type1 == type2` check to see if they are equals for fast check.

Aaaaaaron replied:
> Add `type1 == type2` check to see if they are equals for fast check.

ok

danny0405 commented on core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:
```diff
@@ -117,6 +120,45 @@
+  @Test void testEqualAsCollectionSansNullability() {
+    final SqlTypeFactoryImpl factory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+
```
Try to reuse the `Fixture`.

Aaaaaaron replied:
The basic type is replaced, but nested type in SqlTypeFixture does not meet the requirements(map type/nullability), should I add these to SqlTypeFixture? Or just leave them in this ut.

danny0405 replied:
I would prefer add the types to `SqlTypeFixture` and give it a good name.

Aaaaaaron replied:
> I would prefer add the types to `SqlTypeFixture` and give it a good name.

Thanks for your suggestion, updated already.

Aaaaaaron replied:
> I would prefer add the types to `SqlTypeFixture` and give it a good name.

Thanks for your suggestion, updated already.

danny0405 commented on core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:
```diff
@@ -117,6 +120,34 @@
 
+  @Test void testEqualAsCollectionSansNullability() {
+    final SqlTypeFactoryImpl factory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
```
Use `f.typeFactory` instead.

Aaaaaaron replied:
Thanks for your suggestion, updated already.

danny0405 commented on core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:
```diff
@@ -117,6 +120,34 @@
+  @Test void testEqualAsMapSansNullability() {
+    final SqlTypeFactoryImpl factory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+
```
Use `f.typeFactory` instead.

Aaaaaaron replied:
Thanks for your suggestion, updated already.
